### PR TITLE
Add SHAP tutorial for Transformers

### PR DESCRIPTION
**Pull request by lisa-sousa**
## Summary
Adds a new SHAP-based explainability tutorial for Transformer models, extending the existing SHAP coverage (currently used for Random Forests and CNNs) to transformer architectures.

## Changes
- New notebook(s)/content under `xai-for-transformer/` demonstrating SHAP applied to a transformer model
- Updates to support loading a pretrained transformer model for the tutorial

## ⚠️ Notes for the merger @ema265 

1. **Pin `shap` version in requirements**
   `requirements_xai-for-transformer.txt` needs `shap==0.52.0` pinned. Please also check whether `requirements_xai-for-cnn.txt` and `requirements_xai-for-random-forest.txt` can be pinned to the same version — if so, update them too and re-run those notebooks end-to-end to confirm nothing breaks with the newer SHAP release before merging.

2. **Enable automatic model weights download**
   The transformer notebook currently expects model weights to be available locally/manually placed. Model weights will be published as part of a **GitHub Release** — once that release is live, switch on/uncomment the automatic download step (pointing at the release asset URL) so users don't have to source weights manually.

3. **Double-check reported train/test accuracy for the loaded transformer model**
   When running the intro notebook, I get:
   - TRAIN accuracy: 0.9498
   - TEST accuracy: 0.8776

   using balanced accuracy. Can you confirm whether the notebook's original reported numbers were computed with balanced accuracy or a different metric (plain accuracy, F1, etc.)? If the notebook text states different figures, we should either update the text to match, or check if there's a metric/data-split mismatch causing the discrepancy.

### Testing checklist
- [x] Notebooks in `xai-for-transformer/` run end-to-end with `shap==0.52.0`
- [x] Notebooks in `xai-for-cnn/` and `xai-for-random-forest/` re-tested with pinned SHAP version
- [x] Model weights auto-download verified against the GitHub Release once published
- [x] Accuracy figures in notebook text reconciled with actual output (confirm metric used)

### CODE REVIEW

File: docs/source/_figures/shap_partition_explainer.png
Comment (donatella-cea) [inline docs/source/_figures/shap_partition_explainer.png:1]: Figure feedback: I like the overall workflow, but I find the figure difficult to read at notebook size. The annotations are quite small, and I have to zoom in to read them. A few suggestions:

Panel 2: The hierarchy appears to always have three levels. Since the depth depends on the partition tree, perhaps indicate omitted intermediate levels (e.g., ⋮ or "Level n"). If using tree terminology, labels such as Root (all tokens) and Leaf nodes (individual tokens) may make this clearer.
Panel 4: This panel is quite dense. Consider simplifying some annotations or moving some explanatory text to the figure caption.

Panel 4: I would rename it "Recursively allocate contributions".
The root contribution is +0.62, which matches the first prediction shown in Panel 3. Is this intentional? If so, could the connection be made explicit? Otherwise, using a different value might avoid the impression that the root contribution is copied from the first model evaluation.

Panels 5: In Panel 4, all recursively allocated contributions are shown as positive values, whereas Panel 5 contains both positive (red) and negative (blue) SHAP values. This made me wonder where the negative contributions arise. Perhaps using both positive and negative example values in Panel 4, or explicitly stating that the numbers are illustrative, would make the connection clearer.
I would replace "Obtain" with "Visualize".

File: xai-for-transformer/x-Tutorial_SHAP_Text.ipynb
Comment (donatella-cea) [inline xai-for-transformer/x-Tutorial_SHAP_Text.ipynb:1]: - Number of explained texts --> number of input sentences or texts
- Clarify base value vs base value prediction: Consider adding "For models with multiple outputs, each output has its own base value, corresponding to the model's baseline prediction for that output." at the beginning of the force plot section. At the end of the local explanation I would would use 'baseline prediction' instead of 'base value'.
-typo: acending order
-bar plot mean( ): "average importance across all explained texts," or "total importance across all occurrences." ?

**Review by donatella-cea — COMMENTED** [review]
Really nice and easy to follow notebook!
I have a few suggestions on the figure and some small clarifications.

File: xai-for-transformer/utils.py [hunk 1]
```diff
@@ -122,3 +125,80 @@ def plot_attention_maps(img, attentions):
...
+    # ground-truth labels to compute the balanced classification accuracy.
+    predicted_ids = [label2id[max(p, key=lambda x: x["score"])["label"]] for p in predictions]
+    accuracy = balanced_accuracy_score(data["emotion"], predicted_ids)
+    return accuracy
+
+def plot_shap_values(shap_values_1, shap_values_2, token_name, max_display=20):
+
+    # Aggregate SHAP values
+    train = shap_values_1[:, :, token_name].mean(0)
+    test = shap_values_2[:, :, token_name].mean(0)
```
[the remark this answers is not in the document]
Comment (lisa-sousa) [reply xai-for-transformer/utils.py]: renamed to output_name

File: xai-for-transformer/utils.py [hunk 2]
```diff
@@ -10,6 +10,9 @@
 
 from torchvision import transforms
 
+from sklearn.metrics import balanced_accuracy_score
+
+import shap
```
[the remark this answers is not in the document]
Comment (lisa-sousa) [reply xai-for-transformer/utils.py:15]: This will be fixed once all notebooks will be integrated into the main branch

**Comment by ema265** [comment]
Hi @lisa-sousa, I started on merger note #1. I pinned `shap==0.52.0` in the transformer requirements, then tested it end-to-end, and a few things came up I'd like to align on before changing anything else:

1. **shap 0.52.0 requires Python ≥ 3.12.** On 3.11 the pin makes `pip install -r requirements_xai-for-transformer.txt` fail (the newest shap for 3.11 is 0.51.0). So this effectively moves the transformer track to 3.12, and bumping the CNN/RF requirements to 0.52.0 would force those tracks to 3.12 as well. Are we OK moving them to 3.12?

2. **`requirements_xai-for-transformer.txt` is missing `datasets`.** The notebook imports it (to load `dair-ai/emotion`), so it won't run from the requirements alone. I can add it if that's fine.

3. **token_type_ids crash.** With the pinned `transformers==4.51.3` and the hosted DistilBERT weights, the pipeline fails with `DistilBertForSequenceClassification.forward() got an unexpected keyword argument 'token_type_ids'` (the saved tokenizer lists `token_type_ids`, which DistilBERT doesn't accept). It's fixable by re-saving the tokenizer in the weights Release, or a one-line workaround in the notebook (`tokenizer.model_input_names = ["input_ids", "attention_mask"]`). Which transformers version did you test the notebook with? That would tell me whether it's just a version difference on my side.

The SHAP part itself works fine on 0.52.0 (PartitionExplainer, the text masker, and the plots all run), so the pin is the right version. I just wanted to agree on the Python-3.12 move and the token_type_ids fix before touching CNN/RF. Let me know how you'd like to proceed.

**Comment by ema265** [comment]
On #3 (accuracy check): it's just the metric, not a data/split mismatch. On the same test split I get:

- plain accuracy: **0.9295**
- balanced accuracy: **0.8776** (your number)
- weighted F1: **0.929**

So the 0.9498 / 0.8776 the notebook prints are **balanced accuracy** (that's what `utils.evaluate_model` computes), while the 92.95% in the training summary is **plain accuracy**. Both are correct, same split.

The notebook was labeling them just "accuracy", so I relabeled the prints to "balanced accuracy" and added a short note explaining why we use it (the emotion classes are imbalanced). Pushed to the branch (76711dc).

**Comment by ema265** [comment]
@donatella-cea when you have a moment, could you publish the draft weights Release?

Task #2 on this PR (automatic download of the fine-tuned DistilBERT weights) is ready in the notebook, but it is blocked because the **"DistilBERT emotion weights"** release (tag `distilbert-emotion-weights`) is still a **draft**, so its download URL returns 404. Once it is published the notebook's download step should work as is, since the URL already matches the tag.

Publishing it also unblocks the "Load the fully fine-tuned model" cell in **notebook 1**, so one publish covers both notebooks.

You can find it under the repo's [Releases tab](https://github.com/HelmholtzAI-Consultants-Munich/XAI-Tutorials/releases), named "DistilBERT emotion weights". Thanks!

**Comment by lisa-sousa** [comment]
> @donatella-cea when you have a moment, could you publish the draft weights Release?
> 
> Task #2 on this PR (automatic download of the fine-tuned DistilBERT weights) is ready in the notebook, but it is blocked because the **"DistilBERT emotion weights"** release (tag `distilbert-emotion-weights`) is still a **draft**, so its download URL returns 404. Once it is published the notebook's download step should work as is, since the URL already matches the tag.
> 
> Publishing it also unblocks the "Load the fully fine-tuned model" cell in **notebook 1**, so one publish covers both notebooks.
> 
> You can find it under the repo's [Releases tab](https://github.com/HelmholtzAI-Consultants-Munich/XAI-Tutorials/releases), named "DistilBERT emotion weights". Thanks!

Hi @ema265 I published the model weights as a new release. I'm following our release versioning, hence the new release is v2.1.0: https://github.com/HelmholtzAI-Consultants-Munich/XAI-Tutorials/releases/tag/v2.1.0

**Comment by lisa-sousa** [comment]
> Hi @lisa-sousa, I started on merger note #1. I pinned `shap==0.52.0` in the transformer requirements, then tested it end-to-end, and a few things came up I'd like to align on before changing anything else:
> 
>     1. **shap 0.52.0 requires Python ≥ 3.12.** On 3.11 the pin makes `pip install -r requirements_xai-for-transformer.txt` fail (the newest shap for 3.11 is 0.51.0). So this effectively moves the transformer track to 3.12, and bumping the CNN/RF requirements to 0.52.0 would force those tracks to 3.12 as well. Are we OK moving them to 3.12?
> 
>     2. **`requirements_xai-for-transformer.txt` is missing `datasets`.** The notebook imports it (to load `dair-ai/emotion`), so it won't run from the requirements alone. I can add it if that's fine.
> 
>     3. **token_type_ids crash.** With the pinned `transformers==4.51.3` and the hosted DistilBERT weights, the pipeline fails with `DistilBertForSequenceClassification.forward() got an unexpected keyword argument 'token_type_ids'` (the saved tokenizer lists `token_type_ids`, which DistilBERT doesn't accept). It's fixable by re-saving the tokenizer in the weights Release, or a one-line workaround in the notebook (`tokenizer.model_input_names = ["input_ids", "attention_mask"]`). Which transformers version did you test the notebook with? That would tell me whether it's just a version difference on my side.
> 
> 
> The SHAP part itself works fine on 0.52.0 (PartitionExplainer, the text masker, and the plots all run), so the pin is the right version. I just wanted to agree on the Python-3.12 move and the token_type_ids fix before touching CNN/RF. Let me know how you'd like to proceed.

Hi @ema265 
regarding your questions:
1. Yes, please pin Python version 3.12 for all requirements (RF, CNNs, Transformers), so we have the same Pyhton version everywhere. 
2. Yes, please add. Also in case any other requirement is missing. Please pin versions for all requirements (this can be updated when other notebooks are merged).
3. I used: `transformers==5.13.0`

**Comment by ema265** [comment]
Hi @lisa-sousa, status from the merger side. Thanks for the answers and for publishing the weights.

**#1 pin shap 0.52.0:** done. I pinned shap==0.52.0, bumped transformers to 5.13.0 (that cleared the token_type_ids crash, so it was purely the version, thanks for the pointer), and added the missing deps (datasets, pandas, requests), all pinned. I also added a "Requires Python 3.12" marker to all three requirements files (transformer, CNN, RF) as you asked. I re-tested the SHAP notebook end-to-end on a clean Python 3.12 env: full run, no errors, and it reproduces the published accuracy (TRAIN 0.9498, TEST 0.8776).

**#2 weights auto-download:** done. I pointed the download at the new v2.1.0 release and verified it from a clean state (the notebook downloads, unzips, and loads the model).

**#3 accuracy:** done earlier. The prints are balanced accuracy (from utils.evaluate_model), relabeled with a short note on why we use it.

**Testing checklist:**
- Transformer notebooks with shap 0.52.0: checked. Of the notebooks in the folder only the SHAP one uses shap, and it runs clean on 3.12.
- CNN and RF re-tested with pinned SHAP: I left this unchecked on purpose. Per your note the SHAP version alignment happens once all notebooks are integrated, so I only added the 3.12 marker to their requirements and did not bump or re-run them. Flagging it as your integration item so the empty box is not read as forgotten.
- Weights auto-download and accuracy reconciled: both checked.

**Before I merge:** Donatella left two review threads that are still open, and since they are both on content you authored, you might want to take a look before I merge:
- the SHAP figure (shap_partition_explainer.png) readability notes (panels 2/4/5, "Obtain" to "Visualize", and so on)
- the notebook text tweaks (the "acending" typo, the "number of explained texts" wording, the base value clarification sentence, and the bar plot mean wording)

It would be good to have these resolved first. The figure is best updated on your side, but I am happy to apply the small text fixes myself if you would prefer.

**Comment by lisa-sousa** [comment]
@ema265 thanks for the addressing all those points.

I still have to update the notebook content according to Donatellas comments and I also reworked the full video slides and reda the docs content, so I will tailor the notebook accordingly. I'll let you and Donatella know when I updated the notebook.

File: docs/source/_figures/shap_partition_explainer.png
Comment (lisa-sousa) [reply docs/source/_figures/shap_partition_explainer.png:1]: Removed figure from notebook, because the exhaustive explanation for Partial Explainer moved to read the docs

File: xai-for-transformer/x-Tutorial_SHAP_Text.ipynb
Comment (lisa-sousa) [reply xai-for-transformer/x-Tutorial_SHAP_Text.ipynb:1]: adressed all issues in notebook update

**Comment by lisa-sousa** [comment]
@donatella-cea could you please have a look at the updated transformer notebook? I’ve addressed your comments and aligned the content with the new Read the Docs structure for SHAP.

In addition, could you briefly review the updated SHAP content in the Random Forest and CNN notebooks? These have also been revised to follow the new structure and now include additional explainers: KernelSHAP for the Random Forest and PartitionExplainer for the CNN.

@ema265 once Donatella has approved the changes, this should be ready to merge. Before merging, could you please double-check that all three notebooks still run correctly with the updated requirements? Thanks :)

**Comment by donatella-cea** [comment]
> @donatella-cea could you please have a look at the updated transformer notebook? I’ve addressed your comments and aligned the content with the new Read the Docs structure for SHAP.
> 
> In addition, could you briefly review the updated SHAP content in the Random Forest and CNN notebooks? These have also been revised to follow the new structure and now include additional explainers: KernelSHAP for the Random Forest and PartitionExplainer for the CNN.
> 
> @ema265 once Donatella has approved the changes, this should be ready to merge. Before merging, could you please double-check that all three notebooks still run correctly with the updated requirements? Thanks :)

Everything looks good to me! @ema265, as Lisa mentioned, you can go on with the merging after double-checking that the notebooks run with the updated requirements. Thanks!

**Comment by lisa-sousa** [comment]
@ema265 

Regarding google collab section, could you try out this version:

```
# Move to Colab's temporary working directory
%cd /content

# Remove any existing copy and clone the specified course branch
!rm -rf XAI-Tutorials
!git clone --depth 1 --branch main https://github.com/HelmholtzAI-Consultants-Munich/XAI-Tutorials.git
```

Like this the student should not be ofrced to share their google drive for the course purpose and with --depth 1 it should create a shallow copy of the branch such that less data needs to be downloaded (important for on-site courses where the bandwidth might be liited).

Thank you!

**Comment by ema265** [comment]
@lisa-sousa tested the shallow clone on macOS, and the numbers are good:

```
full clone:   661 MB on disk, 513 MB of that .git, 53.7s
--depth 1:    276 MB on disk, 128 MB of that .git, 15.8s
```

That is 385 MB and about 38 seconds saved. I diffed both working trees recursively to confirm nothing is missing: notebooks, models, data and requirements are identical.

Dropping the Drive mount is a clear improvement too. When I tested the CNN notebooks on Colab recently, the mount was the step that actually failed for me, so removing it takes out a real point of friction.

I have pushed it to the branch, with two adjustments:

**It is guarded so it only runs on Colab.** Your version is not commented out, which is what makes it convenient, but it also means it runs everywhere. Locally `%cd /content` fails, and more seriously `rm -rf XAI-Tutorials` could delete somebody's own clone if they run the notebook from the parent directory. The cell now checks for `google.colab` and does nothing off Colab.

**The three cells are now one.** Mount, clone and install were separate cells, all commented. Since the clone step now changes directory and the install step depended on that, keeping them apart meant a restart between them would break the working directory. They are one cell, so it either all runs or none of it does.

Applied to all 25 notebooks that had the old three-cell setup, so the modules do not drift apart. Happy to split that into its own PR if you would rather keep this one focused on SHAP, it is a single commit and easy to move.

One number worth knowing: even shallow it is 276 MB, because the working tree alone is 147 MB of committed data and weights. Better, but not small on a weak connection. If bandwidth is the real concern for on-site courses, a sparse checkout of just the module being taught would be the bigger lever, though that is a much more invasive change.

**Comment by lisa-sousa** [comment]
Hi @ema265 
Thanks for the review. The two changes you suggested sound good. Regarding the size of the check out repo, most of the time we only subselect content form the full repo, so what we check out would probably be even less than, but good that the shallow clone is working :)

I would suggest the this PR now:

- [x] move the collab changes to a separate PR, so we have an overview where we changed it
- [ ] once you have moved out the collab changes, you can merge this PR inti main

Thank you!

**Comment by ema265** [comment]
Done, the Colab changes are now in #41 and this branch is back to just the SHAP work. I will continue with the second point.
